### PR TITLE
CORE-9451 simplify crypto service configuration handling

### PR DESCRIPTION
**Pull request by dickon**
Previously the crypto code was tightly coupled with the config service. In particular:

1. The crypto service created its own default configuration, with embedded passphrase, and then for the combined worker stored this in its database. (Everything else in Corda uses the Config schema to handle defaults). Special measures were taken to make sure the default master wrapping key passphrase was encrypted, which required knowledge of how smart config handles secrets which is no longer safe to assume since we allow different secrets implementations.
2. Different crypto service providers were supported, each with its own configuration. In order to handle this in a general way, the smart configuration for the crypto service was converted to JSON by CryptoServiceFactoryImpl, then dynamically instantiated into configuration objects particular to a specific crypto service provider, and passed in dynamically typed. This also involved care to make sure that secrets such as the passphrase could still be decrypted, which make assumptions that secrets are always encrypted locally. With other sources of secrets that assumption is no longer sound.
3. The configuration of the crypto service providers was handled by a layer of data classes allowing plain access to the classes, rather than using methods of SmartConfig. That's reasonable of itself, but is different to the rest of Corda. 
4. Some of the crypto code at lib level was configurable, which would normally have caused a dependency on the confgiuration library. However, the configuration library has a dependency on the crypto library, in order to handle cryptogrpahy on the the secret parts of the config. This prevented some parts of the crypto code from using SmartConfig.
5. There is a separate Corda crypto technical compliance kit which shares code with the production implementation.

In combination, this means that the configuratioon of crpto scheme is complex and brittle, and would get significantly more complex when extended to include support for third party secret stores. 

In this change, I've moved to mostly using SmartConfig directly, much like the rest of Corda. I've also moved the library-level generic (templated) crypto service provider, and instead we have the components/crypto/crypto-softhsm-impl provide a simplified, non-templated/non-generic crypto service provide single abstract method interface and a simple implementation which provides a simplified type for the crypto service.  We may well never need the flexbility of different config service providers, and it is expensive to maintain and would have made this change impractical.

Having components/crypto/crypto-softhsm-impl define the CryptoServiceProvider isn't ideal, but:
a) it fits there without introducing a dependency cycle with crypto-service-impl
b) we plan to shortly collapse down the support for multiple different types of HSM, and merge crypto-softhsm-impl and crypto-sevice-impl

The Crypto service at this point still has its own way of constructing default configuration to go into the database. We may get rid of this soon, and instead simply use the JSON config schema as the rest of the Corda does. For now we pass in the passphrae and salt default values.

Extend some comments and a README on how config works based on what I figured out about configuration to design and implement this change. 

Add a toString() to dump out AesKeys in Base64. To be used with care, but if you need to actually compare keys to debug this helps.

Delete KeyCredentials, its tests and methods in other classes with initialise based on it, since we no longer use it. I simply
have the createDefaultCryptoConfig take two parameters instead of needing a dataclass.

### CODE REVIEW

File: processors/member-processor/src/integrationTest/kotlin/net/corda/processor/member/MemberProcessorTestUtils.kt [hunk 1]
```diff
@@ -279,7 +277,7 @@ class MemberProcessorTestUtils {
 
         fun Publisher.publishMembershipConf(membershipConfig: SmartConfig) =
             publishConf(ConfigKeys.MEMBERSHIP_CONFIG, membershipConfig.root().render())
-        fun Publisher.publishDefaultCryptoConf(cryptoConfig: SmartConfig) =
+        fun Publisher.publishDefaultCryptoConf(cryptoConfig: Config) =
```
Comment (driessamyn) [inline processors/member-processor/src/integrationTest/kotlin/net/corda/processor/member/MemberProcessorTestUtils.kt]: I don't think this change is necessary.

**Review by driessamyn — COMMENTED** [review]
Looks good to me, just one minor comment, but will wait for @kyriathar 's review.

File: processors/member-processor/src/integrationTest/kotlin/net/corda/processor/member/MemberProcessorTestUtils.kt (hunk 1 above)
Comment (dickon) [reply processors/member-processor/src/integrationTest/kotlin/net/corda/processor/member/MemberProcessorTestUtils.kt]: Yes, it was left over from an earlier direction I took and then largley reverted. Fixed.

File: applications/workers/release/crypto-worker/src/main/kotlin/net/corda/applications/workers/crypto/CryptoWorker.kt [hunk 2]
```diff
@@ -98,6 +98,7 @@ class CryptoWorkerParams {
     @CommandLine.Option(names = ["-d", "--database-params"], description = ["Database parameters for the worker."])
     var databaseParams = emptyMap<String, String>()
 
+    // TODO - delete me
     @CommandLine.Option(names = ["--hsm-id"], description = ["HSM ID which is handled by this worker instance."])
     var hsmId = ""
```
Comment (kyriathar) [inline applications/workers/release/crypto-worker/src/main/kotlin/net/corda/applications/workers/crypto/CryptoWorker.kt]: I think this refers to the hard HSMs which we will not support so I agree it should be removed. Could we please raise a JIRA ticket and reference it from here if not intended to remove in this PR?

File: applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt [hunk 3]
```diff
@@ -52,17 +52,29 @@ class WorkerHelpers {
             return params
         }
 
+        data class BootstrapConfigResults(
+            val bootstrapConfig: SmartConfig,
+            val smartConfigFactory: SmartConfigFactory
+        )
```
Comment (kyriathar) [inline applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt]: I think `SmartConfigFactory` is already available through `SmartConfig.factory` so maybe we don't need this data class.

File: applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt [hunk 4]
```diff
@@ -52,17 +52,29 @@ class WorkerHelpers {
...
+
         /**
          * Uses [smartConfigFactory] to create a `SmartConfig` containing the instance ID, topic prefix, additional
-         * params in the [defaultParams], and any [extraParams].
+         * params in the [defaultParams], and any [extraParams]. Check that the configuration matches the schema,
+         * but do not fill in defaults.
```
Comment (kyriathar) [inline applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt]: nitpicky: I think there are mentioned some implementation details here. Could we simplify to say: 

"Creates bootstrap configuration out of command line configuration and gets validated against schema: `corda.boot.json`

It is then passed to workers' processors and also to `ConfigurationReadService` (see `ConfigurationReadService.bootstrapConfig`)."

Maybe also briefly say what it contains?

File: applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt [hunk 5]
```diff
@@ -87,13 +99,29 @@ class WorkerHelpers {
...
+            // Also, boot config has been validated; all the keys in it are declared in the JSON schema and are of
+            // the types specified in the schema
+            //
+            // However, bootConfig does not:
+            //  - have database configuration records applied
+            //  - have unspecified fields filled in with defaults from the schema.
+            //
+            // This is handled later, via the lifecycle coordinator and the domino logic,
+            // based on the requirements graph.
+            //
```
Comment (kyriathar) [inline applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt]: I think maybe remove this comment? First part should be mentioned on kDoc. Then the defaulting I think you mean happens later. Here the defaulting is whatever comes in from for `DefaultWorkerParams` and workers specific defaulted fields.

File: applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt [hunk 6]
```diff
@@ -87,13 +99,29 @@ class WorkerHelpers {
             val secretsConfig =
                 ConfigFactory.parseMap(defaultParams.secretsParams.mapKeys { (key, _) -> "${ConfigKeys.SECRETS_CONFIG}.${key.trim()}" })
 
-            val bootConfig = SmartConfigFactory
+            val smartConfigFactory = SmartConfigFactory
```
Comment (kyriathar) [inline applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt:103]: We could remove this variable and inline with `bootConfig`.

File: components/configuration/README.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+`SmartConfig` inherits the basic approach of typesafe config, which is to take data sources (e.g. JSON files) at
+run time, and provide convenient access functions which throw exceptions if the requested configuration is
+missing, or has the wrong type. This makes it concise and convenient to pass around `SmartConfig` instances and
+look up data on them where needed, without needing explicit type conversion or error checking code. This approach
+to configuration is dynamic; neither the IDE nor the compiler can tell if the wrong name or type is used. It
+feels more like Python than Java.
+
+Corda is dynamic enough that the configuration needs to be dynamic as well. For instance, in some situations,
+Corda can create new database credentials which we store in the configuration system. So as well as external
+soruces of configuration, we also write configuration records to a to Kafka so that they can be passed around multiple
```
Comment (kyriathar) [inline components/configuration/README.md]:
```suggestion
sources of configuration, we also write configuration records to Kafka so that they can be passed around multiple
```

File: components/configuration/README.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+Typically, this is the sequence for initialisation of configuration:
+
+1. A worker starts as a top level JVM command line program, via OSGi, and gets its command line arguments.
+   This might be a single-purpose worker such as the RPC worker, or, in the non-production local development mode  
+   might be the combined worker.
+2. `WorkerHelpers.getBootstrapConfig` is called to put together a validated smart config object. This
+   bootstrap configuration will typically be incomplete since it will only contain keys specified on the command line,
+   rather than all keys specified in the JSON schema of the configuration in the `corda-api.git/data/config-schema`,
+   but what is in it is guaranteed to match the schema, or else this step will fail.
```
Comment (kyriathar) [inline components/configuration/README.md]: I believe when you say "all keys specified" you mean dynamic config coming in from RPC. So different than bootstrap config. I think this section might be a bit misleading.

File: components/configuration/README.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
+Corda makes use of the [typesafe config](https://github.com/lightbend/config) configuration library. We extend
```
Comment (kyriathar) [inline components/configuration/README.md]: I think let's remove this README file from here and lets make it a confluence wiki page. Then we can review there and potentially add it back here. I think, it is currently very detailed making it a bit hard to follow. Also might become obsolete if any of the steps changes.

File: components/configuration/configuration-read-service/src/main/kotlin/net/corda/configuration/read/ConfigurationReadService.kt [hunk 10]
```diff
@@ -67,12 +67,14 @@ interface ConfigurationReadService : Lifecycle {
...
+     * the BootstrapConfigProvided event is processed. However, calling this multiple times with identical
+     * bootstrap configurations is harmless (and is simply logged at debug level).
+     * An error is also thrown later if the configuration does not allow the configuration read service to
+     * access the full configuration.
      *
-     * @param config The bootstrap configuration to connect to the message bus.
-     * @throws ConfigurationException If the bootstrap configuration is provided after a connection has been
-     *                                established, or if the configuration does not allow access.
+     *  @param config The bootstrap configuration to connect to the message bus.
+     *  @throws ConfigurationException If the bootstrap configuration is provided after a connection has been
```
Comment (kyriathar) [inline components/configuration/configuration-read-service/src/main/kotlin/net/corda/configuration/read/ConfigurationReadService.kt]: I would revert these kDocs.

File: applications/workers/release/combined-worker/src/main/kotlin/net/corda/applications/workers/combined/CombinedWorker.kt [hunk 11]
```diff
@@ -94,7 +94,16 @@ class CombinedWorker @Activate constructor(
...
         val cryptoConfig = PathAndConfig(BOOT_CRYPTO, createCryptoBootstrapParamsMap(params.hsmId))
-        val config = getBootstrapConfig(
+
+        // getBootstrapConfig involves setting up a smart config factory, in order to provide access to
+        // the boot configuration. So, to prevent having to set up smart config factory in several places,
+        // we have getBootstrapConfig return the smart config factory as well as a smart config instance representing
+        // the boot config when we get hold of the boot config.
+        //
+        // It may be worth separating out the smart config factory setup so that we can pass it in to both
+        // getBootstrapConfig and the PostgresDbSetup below.
```
Comment (kyriathar) [inline applications/workers/release/combined-worker/src/main/kotlin/net/corda/applications/workers/combined/CombinedWorker.kt]: Should we remove this comment?

File: libs/crypto/crypto-config-impl/src/main/kotlin/net/corda/crypto/config/impl/CryptoConfigUtils.kt [hunk 12]
```diff
@@ -160,7 +156,7 @@ private const val CRYPTO_CONNECTION_FACTORY_OBJ = "cryptoConnectionFactory"
 private const val SIGNING_SERVICE_OBJ = "signingService"
 private const val HSM_SERVICE_OBJ = "hsmService"
 private const val HSM_MAP = "hsmMap"
-private const val HSM_MAP_ITEM_OBJ = "hsmMap.%s"
+const val HSM_MAP_ITEM_OBJ = "hsmMap.%s"
```
Comment (kyriathar) [inline libs/crypto/crypto-config-impl/src/main/kotlin/net/corda/crypto/config/impl/CryptoConfigUtils.kt]: I think revert back to private? Seems to only getting used in here.

File: tools/plugins/initial-config/src/main/kotlin/net/corda/cli/plugin/initialconfig/CryptoConfigSubcommand.kt [hunk 13]
```diff
@@ -59,25 +60,27 @@ class CryptoConfigSubcommand : Runnable {
...
-                }
-            )
-        ).root().render(ConfigRenderOptions.concise())
+        val smartConfigFactory = SmartConfigFactory.createWith(
+            ConfigFactory.parseString("""
+            ${EncryptionSecretsServiceFactory.SECRET_PASSPHRASE_KEY}=${passphrase}
+            ${EncryptionSecretsServiceFactory.SECRET_SALT_KEY}=${salt}
+        """.trimIndent()
+            ),
+            listOf(EncryptionSecretsServiceFactory())
```
Comment (kyriathar) [inline tools/plugins/initial-config/src/main/kotlin/net/corda/cli/plugin/initialconfig/CryptoConfigSubcommand.kt:72]: My understanding is, in this cli tool we statically use `EncryptionSecretsServiceFactory`. But Corda could be using some different one? In which case it will not be able to decrypt these secrets?

File: libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/aes/KeyCredentials.kt [hunk 14]
```diff
@@ -1,18 +0,0 @@
-package net.corda.crypto.core.aes
```
Comment (kyriathar) [inline libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/aes/KeyCredentials.kt:1]: Removing this class has increased the changes footprint of this PR I think.

**Review by kyriathar — COMMENTED** [review]
Some comments, still need to look at dynamic config that we removed.

File: applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt (hunk 3 above)
Comment (driessamyn) [reply applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt]: Good spot. We don't need this I think.

File: applications/workers/release/crypto-worker/src/main/kotlin/net/corda/applications/workers/crypto/CryptoWorker.kt (hunk 2 above)
Comment (dickon) [reply applications/workers/release/crypto-worker/src/main/kotlin/net/corda/applications/workers/crypto/CryptoWorker.kt]: I don't think we need the overhead of a JIRA ticket for each tech debt TODO. I'll clarify the comment though, it's too vague.

File: components/configuration/README.md (hunk 9 above)
Comment (driessamyn) [reply components/configuration/README.md]: I think some of this is already captured in the github wiki, so may you're right, maybe it should go there.

File: applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt (hunk 3 above)
Comment (dickon) [reply applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt]: Yes, nice, done.

File: applications/workers/release/crypto-worker/src/main/kotlin/net/corda/applications/workers/crypto/CryptoWorker.kt (hunk 2 above)
Comment (dickon) [reply applications/workers/release/crypto-worker/src/main/kotlin/net/corda/applications/workers/crypto/CryptoWorker.kt]: Jira Link added

File: applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt (hunk 4 above)
Comment (dickon) [reply applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt]: The comment does address is implementation details. I don't see any problem with having such material in this comment, nor do I see anywhere else to put such a comment.  If this was split out to an interface and a subclass that implements the interface then you'd have a point. I did take the rest of your comment.

File: applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt (hunk 5 above)
Comment (dickon) [reply applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt]: I updated the comment to clarify a little.

File: applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt (hunk 6 above)
Comment (dickon) [reply applications/workers/worker-common/src/main/kotlin/net/corda/applications/workers/workercommon/WorkerHelpers.kt:103]: This changed a bit as a result of using SmartConfig.factory; I think it is okay now.

File: components/configuration/README.md (hunk 8 above)
Comment (dickon) [reply components/configuration/README.md]: I don't think it's misleading, and I'm not sure what you are getting at.

File: components/configuration/configuration-read-service/src/main/kotlin/net/corda/configuration/read/ConfigurationReadService.kt (hunk 10 above)
Comment (dickon) [reply components/configuration/configuration-read-service/src/main/kotlin/net/corda/configuration/read/ConfigurationReadService.kt]: :+1: done

File: components/configuration/README.md (hunk 9 above)
Comment (dickon) [reply components/configuration/README.md]: It's not clear that there's any reason to keep this internal. I agree it is tightly coupled to the code, which was one reason I put it here. I'm not sure how the github wiki handles branching, but usually wikis aren't good at that,  so the reason we still have significant detail in the READMEs is to give us the opportunity and encouragement to keep it in sync with the code.  Github wikis are good for point-in-time text. 

I don't think any of the detail here is irrelevant (and feel free to quibble on particular clauses in the text if you disagree) if you want to understand how configuration works, and configuration is important.  Simplification would be welcome, although I don't see much opportunity given our requirements.

File: applications/workers/release/combined-worker/src/main/kotlin/net/corda/applications/workers/combined/CombinedWorker.kt (hunk 11 above)
Comment (dickon) [reply applications/workers/release/combined-worker/src/main/kotlin/net/corda/applications/workers/combined/CombinedWorker.kt]: Why?

File: libs/crypto/crypto-config-impl/src/main/kotlin/net/corda/crypto/config/impl/CryptoConfigUtils.kt (hunk 12 above)
Comment (dickon) [reply libs/crypto/crypto-config-impl/src/main/kotlin/net/corda/crypto/config/impl/CryptoConfigUtils.kt]: yeah done

File: tools/plugins/initial-config/src/main/kotlin/net/corda/cli/plugin/initialconfig/CryptoConfigSubcommand.kt (hunk 13 above)
Comment (dickon) [reply tools/plugins/initial-config/src/main/kotlin/net/corda/cli/plugin/initialconfig/CryptoConfigSubcommand.kt:72]: Yeah, comment and ticket added.

File: libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/aes/KeyCredentials.kt (hunk 14 above)
Comment (dickon) [reply libs/crypto/crypto-core/src/main/kotlin/net/corda/crypto/core/aes/KeyCredentials.kt:1]: Right, but it's also dead code, since it was only used in this one place, so to not do so would have increased tech debt.

File: components/configuration/README.md (hunk 7 above)
Comment (dickon) [reply components/configuration/README.md]: done, thanks

File: .run/Combined Worker Local.run.xml [hunk 15]
```diff
@@ -2,11 +2,12 @@
   <configuration default="false" name="Combined Worker Local" type="JarApplication">
     <option name="JAR_PATH" value="$PROJECT_DIR$/applications/workers/release/combined-worker/build/bin/corda-combined-worker-5.0.0.0-SNAPSHOT.jar" />
     <option name="VM_PARAMETERS" value="-Dco.paralleluniverse.fibers.verifyInstrumentation=true" />
-    <option name="PROGRAM_PARAMETERS" value="--instance-id=0 -mbus.busType=DATABASE -spassphrase=password -ssalt=salt -spassphrase=password -ssalt=salt -ddatabase.user=user -ddatabase.pass=password -ddatabase.jdbc.url=jdbc:postgresql://localhost:5432/cordacluster -ddatabase.jdbc.directory=$ProjectFileDir$/applications/workers/release/combined-worker/drivers" />
+    <option name="PROGRAM_PARAMETERS" value="--instance-id=0 -mbus.busType=DATABASE -ddatabase.user=user -ddatabase.pass=password -ddatabase.jdbc.url=jdbc:postgresql://localhost:5432/cordacluster -ddatabase.jdbc.directory=$ProjectFileDir$/applications/workers/release/combined-worker/drivers -p 7004 -stype=hashicorp-vault -surl=http://127.0.0.1:8200 -stoken=myroot -daddon.directory=$PROJECT_DIR$/../corda5-enterprise/addons/configuration/hashicorp-vault/build/libs" />
```
Comment (driessamyn) [inline .run/Combined Worker Local.run.xml]: This should not be committed.

**Review by driessamyn — CHANGES_REQUESTED** [review]

File: applications/workers/release/combined-worker/src/main/kotlin/net/corda/applications/workers/combined/CombinedWorker.kt (hunk 11 above)
Comment (driessamyn) [reply applications/workers/release/combined-worker/src/main/kotlin/net/corda/applications/workers/combined/CombinedWorker.kt]: Is it still valid as the factory is available as a property on the SmartConfig?

File: components/configuration/README.md (hunk 9 above)
Comment (driessamyn) [reply components/configuration/README.md]: The Github wiki isn't internal.

I think the question is really about how many places we have developer documentation in.

I agree it doesn't feel right to be here as it is inconsistent with other documentation.
I feel like it's too  much detail as well, which will go out of date quickly, but I appreciate that's subjective.

File: .run/Combined Worker Local.run.xml (hunk 15 above)
Comment (dickon) [reply .run/Combined Worker Local.run.xml]: Yes, that was probabl a mistake in git rebasing, glad you spotted it.

Comment (dickon) [reply .run/Combined Worker Local.run.xml]: reverted

**Review by ronanbrowne — DISMISSED** [review]
LGTM - build

**Review by driessamyn — APPROVED** [review]
LGTM

**Review by Tom-Fitzpatrick — APPROVED** [review]

**Review by mag1c-48 — APPROVED** [review]
lgtm - build
